x/bep3/genesis.go

```diff
@@ -0,0 +1,64 @@
+package bep3
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/kava-labs/kava/x/bep3/types"
+)
+
+// InitGenesis initializes the store state from a genesis state.
+func InitGenesis(ctx sdk.Context, keeper Keeper, supplyKeeper types.SupplyKeeper, gs GenesisState) {
+	if err := gs.Validate(); err != nil {
+		panic(fmt.Sprintf("failed to validate %s genesis state: %s", ModuleName, err))
+	}
+
+	totalAssetSupplyCoins := sdk.NewCoins()
+	for _, asset := range gs.Assets {
+		keeper.SetAssetSupply(ctx, asset, []byte(asset.Denom))
+		totalAssetSupplyCoins.Add(sdk.NewCoins(asset))
+	}
+
+	totalAtomicSwapCoins := sdk.NewCoins()
+	for _, genSwap := range gs.AtomicSwaps {
+		swap, ok := genSwap.(types.AtomicSwap)
+		if !ok {
+			panic("could not convert stored GenesisAtomicSwap to AtomicSwap type")
+		}
+		keeper.SetAtomicSwap(ctx, swap, swap.GetSwapID())
+		totalAtomicSwapCoins.Add(swap.GetModuleAccountCoins())
+	}
+
+	keeper.SetParams(ctx, gs.Params)
+
+	totalGenesisCoins := totalAssetSupplyCoins.Add(totalAtomicSwapCoins)
+
+	// Check module coins match expected genesis coins
+	moduleAcc := supplyKeeper.GetModuleAccount(ctx, ModuleName)
+	if !moduleAcc.GetCoins().IsEqual(totalGenesisCoins) {
+		panic(fmt.Sprintf("total coins (%s) do not equal (%s) module account (%s) ", moduleAcc.GetCoins(), ModuleName, totalGenesisCoins))
+	}
+}
+
+// ExportGenesis writes the current store values
+// to a genesis file, which can be imported again
+// with InitGenesis
+func ExportGenesis(ctx sdk.Context, k Keeper) (data types.GenesisState) {
+	params := k.GetParams(ctx)
+
+	genAssetSupplies := []sdk.Coin{}
+	// TODO: Add k.IterateAssetSupplies
+
+	genAtomicSwaps := types.GenesisAtomicSwaps{}
```

Comment:
I think it would be safe to follow the CDP module pattern and just marshal/unmarshall all swaps at the time of InitGenesis/Export Genesis. Because expiry is checked at each state transition, importing expired swaps doesn't seem to pose any risk. In that case, `GenesisState` would look like. 
```
type Swaps []Swap
type GenesisState struct {
	Params      Params             `json:"params" yaml:"params"`
	AtomicSwaps Swaps `json:"atomic_swaps" yaml:"atomic_swaps"`
	Assets      []sdk.Coin         `json:"assets" yaml:"assets"`
}
```

Reply:
Updated in https://github.com/Kava-Labs/kava/pull/370/commits/61daca47e743c4472c8c30a47d575c2cb89af591.
- Is a 'PreviousBlockHeight' required, or is that maintained by the blockchain already?
- Is InitGenesis the best place to set all AssetSupplies to `sdk.NewCoin(denom, sdk.NewInt(0))`? I'm setting each AssetSupply to 0 in CreateAtomicSwap() as they're witnessed, which feels like the wrong way to approach this.

Reply:
Discussed with @rhuairahrighairidh - we agree that our options are:
- leaving the code as is i.e. set an asset's supply the first time it is witnessed
- adding asset supply to params and requiring that asset supply params are always 0 in InitGenesis, then using them to set asset supplies in the store; upon chain restart, the exported asset supplies in the genesis state are used to update the supply from 0 to the actual count.